Document Hint Variables/Constants Opaque/Transparent : rewrite (#19878)
coq/coq: doc/sphinx/addendum/generalized-rewriting.rst

   .. coqtop:: reset in

      Require Setoid.
      Definition f x y := 2*x + y.
      Definition g x y := 2*x + y.

      Goal forall
        (double_f : forall x y, 2*f x y = f (2*x) y + y),
        2 * g 1 8 = 20.
      Proof.

   .. coqtop:: all

        intros.
          assert_succeeds (setoid_rewrite double_f).
          assert_succeeds (rewrite_strat bottomup double_f).
        set (x := g _).
          assert_fails (setoid_rewrite double_f).
          assert_fails (rewrite_strat bottomup double_f).
        Hint Variables Transparent : rewrite.
          assert_succeeds (setoid_rewrite double_f).
          assert_succeeds (rewrite_strat bottomup double_f).
        Hint Constants Opaque : rewrite.
          assert_fails (setoid_rewrite double_f).
          assert_fails (rewrite_strat bottomup double_f).
        subst x.
          assert_fails (setoid_rewrite double_f).
          assert_fails (rewrite_strat bottomup double_f).
        Hint Transparent f g : rewrite.
          assert_succeeds (setoid_rewrite double_f).
          assert_succeeds (rewrite_strat bottomup double_f).

   .. coqtop:: none

        exact eq_refl.
      Qed.

> Constant unfolding during Proper-instance search

Reviewer: ```suggestion
Constant unfolding during `Proper` instance search
```

I think the ``` `Proper` ``` works in section titles but not certain.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  :n:`Hint {| Opaque | Transparent } {+ @qualid } : typeclass_instances.`

``Proper`` instances are resolved using typeclass search. By default, all

> constants are treated as transparent. This may slow down the resolution due to
> a lot of unifications (all the declared ``Proper`` instances are tried at each
> node of the search tree). To speed up the search, declare your non-abbreviation

Reviewer: ```suggestion
constants are treated as transparent. This may slow down the resolution because
:tacn:`typeclasses eauto` will do a lot of unifications (all the declared ``Proper``
instances are tried at each node of the proof search tree).
To speed up the search, declare your non-abbreviation
```

example of a mostly user-space extension of the algorithm.

Constant unfolding during rewriting
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

By default, :tacn:`setoid_rewrite` and :tacn:`rewrite_strat` unfold global
definitions during rewrite rule matching, but do not unfold local
definitions. Unfolding definitions may slow down matching, whereas keeping
definitions opaque may cause matches to be missed.  This behavior is

> configurable using transparency hints in the hint database ``rewrite``:

Reviewer: FWIW: Do you think the variations in how transparent and opaque work for various tactics makes sense or just sort of happened?  If we think it's the latter, making the behavior more consistent may break current projects, so probably not something we will take on :-(

Author: I'm not sure. The process I imagine might have led to the current state of affairs is one of trying to provide some configurability without designing and implementing completely new mechanisms. There clearly is some understanding that this ought to be improved on, as the Ltac2 reduction tactics already take options as first-class arguments. As for what to do with existing code, I am cautiously optimistic: for example, a substantial part of stdlib and CI builds with `Hint Constants Opaque : rewrite` (which I wonder might be a less surprising default).

Reviewer 2: `Hint Constants Opaque : rewrite / typeclass_instances` ought to be there in a GoodDefaults.v file certainly @thomas-lamiaux

Reviewer 3: I suggest we work on it at next CUDW if we haven't by then

Constant unfolding
~~~~~~~~~~~~~~~~~~

.. cmd::  Hint {| Variables | Constants } {| Opaque | Transparent } : rewrite.

   By default, :tacn:`setoid_rewrite` and :tacn:`rewrite_strat` unfold global

>    :term:`definitions <constant>` during rewrite-rule matching, but do not

Author: Indeed, was only moving the existing link.

   unfold local definitions. The former may slow down matching, whereas the

>    latter may cause matches to be missed. This behavior is configurable using

Author: :thinking: 
"the latter" refers to the choice to not unfold local definitions, not the tactic rewrite_strat. It feels weird to say that the choice misses something.

>    latter may cause matches to be missed. This behavior is configurable using

Reviewer: "latter" is ambiguous.  I took it to refer to the last alternatives (unfold or not unfold).  Maybe replace "latter" with what it refers to (rewrite_strat or not unfolding).

>    transparency hints in the special hint database ``rewrite`` as shown,

Reviewer 4: The `rewrite` and `setoid_rewrite` and `rewrite_strat` tactics are hardwired to use the database called `rewrite`.  It is special because its name is hardcoded in the ML source code, nothing else.

>    transparency hints in the special hint database ``rewrite`` as shown,

Author: Indeed, I think it is important to communicate that this rewrite database behaves differently from user-created databases.

>    transparency hints in the special hint database ``rewrite`` as shown,

Reviewer: Yes.  But just saying "special" leaves the reader wondering just how it is special.  And IIUC that it doesn't work for other hint databases.

>    transparency hints in the special hint database ``rewrite`` as shown,

Reviewer 4: > this rewrite database behaves differently from user-created databases.

It's not the database that behaves differently, though, right?  AIUI, there are three "special" databases:
- core - the default db for (e)auto, trivial, etc
- typeclass_instances - the default db for typeclasses eauto
- rewrite - the default db for transparency information for rewrite & setoid_rewrite & rewrite_strat (and autorewrite?), which (unfortunately?) do not allow choosing a different database for transparency information

>    transparency hints in the special hint database ``rewrite`` as shown,

Reviewer 4: @jfehrle what is "it" in "IIUC it doesn't work for other hint databases."?  I believe these commands do work and are meaningful for other hint databases (they should control unification / matching heuristics in (e)auto), it is just that the rewrite database is the one that gets picked up by the rewrite tactics

>    similarly to the effect of :cmd:`Hint Transparent` on unification used by

Reviewer: ```suggestion
   similar to the effect of :cmd:`Hint Transparent` on unification used by
```

>    :tacn:`eauto` and related tactics.

Author: ```suggestion
   :tacn:`eauto` and related tactics. Note that the transparency information from database `rewrite` is used even when rewriting with individual lemmas.
```

   .. coqtop:: reset in

      Require Setoid.
      Definition f x y := 2*x + y.
      Definition g x y := 2*x + y.

      Goal forall
        (double_f : forall x y, 2*f x y = f (2*x) y + y),
        2 * g 1 8 = 20.
      Proof.
        intros.
          assert_succeeds (setoid_rewrite double_f).
          assert_succeeds (rewrite_strat bottomup double_f).

>         set (x :=g _).

Reviewer: maybe show the context here, too

          assert_fails (setoid_rewrite double_f).
          assert_fails (rewrite_strat bottomup double_f).
        Hint Variables Transparent : rewrite.
          assert_succeeds (setoid_rewrite double_f).
          assert_succeeds (rewrite_strat bottomup double_f).
        Hint Constants Opaque : rewrite.
          assert_fails (setoid_rewrite double_f).
          assert_fails (rewrite_strat bottomup double_f).

>         subst x.

Reviewer: ```suggestion
        subst x.   (* uses setoid_rewrite internaly *)
```

Guessing here.

          assert_fails (setoid_rewrite double_f).
          assert_fails (rewrite_strat bottomup double_f).
        exact eq_refl.
      Qed.

> The resolution of ``Proper`` instances is based on :tacn:`typeclasses eauto`

Reviewer: "based on" or "uses"?  The latter is better if true.

Author: I am not sure which way the factoring is, but my sense has been that both call the same underlying machinery.

and hence regards user-defined :term:`constants <constant>` as transparent by
default. This may slow down the resolution due to a lot of unifications (all
the declared ``Proper`` instances are tried at each node of the search tree).
To speed it up,

> declare your constant as rigid for proof search using the command

Reviewer: If "rigid" means "opaque", please use that.  Otherwise I don't know what it means.

> :cmd:`Typeclasses Opaque`.

Author: I don't know.
